server/route.go

```diff
@@ -1,26 +1,110 @@
 package server
 
 import (
+	"errors"
 	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
 	"github.com/smancke/guble/protocol"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+var (
+	errEmptyQueue = errors.New("Empty queue")
+	errTimeout    = errors.New("Channel sending timeout")
+
+	defaultQueueCap = 50
 )
 
+// Route represents a topic for subscription that has a channel to receive messages.
+type Route struct {
+	messagesC chan *MessageForRoute
+	// queue that will store the messages in correct order
+	//
+	// The queue can have a settable size and if it reaches the capacity the
+	// route is closed
+	queue     *queue
+	queueSize int
+
+	// Timeout to define how long to wait for the message to be read on the channel
+	// if timeout is reached the route is closed
+	timeout time.Duration
+	closeC  chan struct{}
+
+	// Indicates if the consumer go routine is running
+	consuming bool
+	invalid   bool
+	mu        sync.RWMutex
+
+	Path          protocol.Path
+	UserID        string // UserID that subscribed or pushes messages to the router
+	ApplicationID string
+	logger        *log.Entry
+}
+
 // NewRoute creates a new route pointer
-func NewRoute(path, applicationID, userID string, channel chan *MessageForRoute) *Route {
-	return &Route{
+func NewRoute(path, applicationID, userID string, c chan *MessageForRoute) *Route {
+	// TODO: this will be received instead of external channel
+	queueSize := 0
+
+	route := &Route{
+		messagesC:     c,
+		queue:         newQueue(queueSize),
+		queueSize:     queueSize,
+		timeout:       -1,
+		closeC:        make(chan struct{}),
 		Path:          protocol.Path(path),
-		messagesC:     channel,
 		UserID:        userID,
 		ApplicationID: applicationID,
+
+		logger: logger.WithFields(log.Fields{
+			"path":          path,
+			"applicationID": applicationID,
+			"userID":        userID,
+		}),
 	}
+	return route
 }
 
-// Route represents a topic for subscription that has a channel to receive messages.
-type Route struct {
-	Path          protocol.Path
-	messagesC     chan *MessageForRoute
-	UserID        string // UserID that subscribed or pushes messages to the router
-	ApplicationID string
+// SetTimeout sets the timeout duration that the route will wait before it will close a
+// blocking channel
+func (r *Route) SetTimeout(timeout time.Duration) *Route {
```

Comment:
To think about: should these 2 properties (timeout + size) be immutable for the Route, or can these be changed safely after Route creation ?
If they are actually immutable in our current usage (also for safe-usage reason), I suggest that we set them just once at Route creation time